go/libraries/doltcore/schema/typeinfo/typeinfo.go

```diff
@@ -0,0 +1,323 @@
+	// ConvertNomsValueToValue converts a Noms value to a go value. The expected NomsKind of the given
+	// parameter is equivalent to the NomsKind returned by this type info.
+	ConvertNomsValueToValue(v types.Value) (interface{}, error)
+
+	// ConvertValueToNomsValue converts a go value or Noms value to a Noms value. The type of the Noms
+	// value will be equivalent to the NomsKind returned from NomsKind.
+	ConvertValueToNomsValue(v interface{}) (types.Value, error)
+
+	// Equals returns whether the given TypeInfo is equivalent to this TypeInfo.
+	Equals(other TypeInfo) bool
+
+	// FormatValue returns the stringified version of the value.
+	FormatValue(v types.Value) (*string, error)
+
+	// GetTypeIdentifier returns an identifier for this type used for serialization.
+	GetTypeIdentifier() Identifier
+
+	// GetTypeParams returns a map[string]string containing the type parameters.  This is used for
+	// serialization and deserialization of type information.
+	GetTypeParams() map[string]string
+
+	// IsValid takes in a value (go or Noms) and returns whether the value is valid for this type.
+	IsValid(v types.Value) bool
+
+	// NomsKind returns the NomsKind that best matches this TypeInfo.
+	NomsKind() types.NomsKind
+
+	// ParseValue parses a string and returns a go value that represents it according to this type.
+	ParseValue(str *string) (types.Value, error)
+
+	// ToSqlType returns the TypeInfo as a sql.Type. If an exact match is able to be made then that is
+	// the one returned, otherwise the sql.Type is the closest match possible.
+	ToSqlType() sql.Type
+
+	// Stringer results are used to inform users of the constraint's properties.
+	fmt.Stringer
+}
+
+// FromSqlType takes in a sql.Type and returns the most relevant TypeInfo.
+func FromSqlType(sqlType sql.Type) (TypeInfo, error) {
+	switch sqlType.Type() {
+	case sqltypes.Null:
+		return UnknownType, nil
+	case sqltypes.Int8:
+		return Int8Type, nil
+	case sqltypes.Int16:
+		return Int16Type, nil
+	case sqltypes.Int24:
+		return Int24Type, nil
+	case sqltypes.Int32:
+		return Int32Type, nil
+	case sqltypes.Int64:
+		return Int64Type, nil
+	case sqltypes.Uint8:
+		return Uint8Type, nil
+	case sqltypes.Uint16:
+		return Uint16Type, nil
+	case sqltypes.Uint24:
+		return Uint24Type, nil
+	case sqltypes.Uint32:
+		return Uint32Type, nil
+	case sqltypes.Uint64:
+		return Uint64Type, nil
+	case sqltypes.Float32:
+		return Float32Type, nil
+	case sqltypes.Float64:
+		return Float64Type, nil
+	case sqltypes.Timestamp:
+		return TimestampType, nil
+	case sqltypes.Date:
+		return DateType, nil
+	case sqltypes.Time:
+		//TODO: determine the storage format
+		if fmt.Sprintf("a") != "" { // always evaluates to true, compiler won't complain about unreachable code
+			return nil, fmt.Errorf(`"%v" has not yet been implemented`, sqlType.String())
+		}
+		return TimeType, nil
+	case sqltypes.Datetime:
+		return DatetimeType, nil
+	case sqltypes.Year:
+		return YearType, nil
+	case sqltypes.Decimal:
```

Comment:
Is this the only code path where we could end up creating one of these types? Really don't want the chance of these ending up in storage until we nail them down.